lib/codec_choose.c

```diff
@@ -90,18 +127,41 @@ codec_choose_arm (struct codec *codec)
 static int
 codec_choose_x86 (struct codec *codec)
 {
-#if (__x86_64__ || __i386__) && (HAVE_AVX2 || HAVE_SSSE3)
+#if (__x86_64__ || __i386__ || _M_X86 || _M_X64 ) && (HAVE_AVX2 || HAVE_SSSE3)
 
 	unsigned int eax, ebx = 0, ecx = 0, edx;
-	unsigned int max_level = __get_cpuid_max(0, NULL);
+	unsigned int max_level;
+
+	#ifdef _MSC_VER 
+	int info[4];
+	__cpuidex(info, 0, 0);
+	max_level = info[0];
+	#else
+	max_level = __get_cpuid_max(0, NULL);
+	#endif
 
 	#if HAVE_AVX2
-	/* Check for AVX2 support: */
+	/* Check for AVX2 support: 
+	 Checking for AVX requires 3 things:
+	 1) CPUID indicates that the OS uses XSAVE and XRSTORE
+	     instructions (allowing saving YMM registers on context
+	     switch)
+	 2) CPUID indicates support for AVX
+	 3) XGETBV indicates the AVX registers will be saved and
+	     restored on context switch
+	
+	 Note that XGETBV is only available on 686 or later CPUs, so
+	 the instruction needs to be conditionally run.*/
 	if (max_level >= 7) {
 		__cpuid_count(7, 0, eax, ebx, ecx, edx);
-		if (ebx & bit_AVX2) {
-			codec->enc = base64_stream_encode_avx2;
-			codec->dec = base64_stream_decode_avx2;
+
+		if ( (ebx & bit_AVX2) && (ecx & bit_XSAVE_XRSTORE)) {
+			unsigned long long xcrFeatureMask;
```

Comment:
The type `unsigned long long` causes a warning on Linux:

```sh
lib/codec_choose.c:159:18: warning: ISO C90 does not support 'long long' [-Wlong-long]
    unsigned long long xcrFeatureMask;
```
I don't understand why this variable should be an `unsigned long long` in the first place, when `_xgetbv_32()` returns an `unsigned long`.

Could you change this variable's type and the return type of `_xgetbv_32()` to `uint32_t`? That will also require including `stdint.h` at the top of the file.

I have some problems with the mixedCasing of the variable name: this code base uses lower_cased_spacing as a convention.